### PR TITLE
feat(function): Support Semi-structured function GET/GET_IGNORE_CASE/GET_PATH

### LINKED ISSUE #4665: Feature: Semi-structured data function get/get_ignore_case/get_path
**Summary**

Add support Semi-structured Data Function `GET`, `GET_IGNORE_CASE`, `GET_PATH`  to get the value from Object or Array.

https://docs.snowflake.com/en/sql-reference/functions/get.html
https://docs.snowflake.com/en/sql-reference/functions/get_ignore_case.html
https://docs.snowflake.com/en/sql-reference/functions/get_path.html

### PR DESCRIPTION
**Pull request by b41sh**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Support `GET` function
Support `GET_IGNORE_CASE` function
Support `GET_PATH` function

## Changelog

- New Feature
- Build/Testing/CI
- Documentation

## Related Issues

Fixes #4665

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

**Comment by BohuTANG** [comment]
@mergify update

File: common/functions/src/scalars/semi_structureds/get.rs [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+            display_name: display_name.to_string(),
+        }))
+    }
+
+    pub fn desc() -> FunctionDescription {
+        FunctionDescription::creator(Box::new(Self::try_create)).features(
+            FunctionFeatures::default()
+                .deterministic()
+                .monotonicity()
+                .disable_passthrough_null()
```
Comment (sundy-li) [inline common/functions/src/scalars/semi_structureds/get.rs:55]: Why not passthrough_null ?

**Comment by BohuTANG** [comment]
```


SyntaxError: /vercel/path0/docs/doc/30-reference/20-functions/110-semi-structured-functions/get_path.md: Expected corresponding JSX closing tag for <index>. (10:409)
--
10:23:22.794 | 8 \| <p>{`Extracts value from an `}<inlineCode parentName="p">{`ARRAY`}</inlineCode>{`, an `}<inlineCode parentName="p">{`OBJECT`}</inlineCode>{`, or a `}<inlineCode parentName="p">{`VARIANT`}</inlineCode>{` by `}<inlineCode parentName="p">{`path_name`}</inlineCode>{`.
10:23:22.794 | 9 \| The value is returned as a `}<inlineCode parentName="p">{`Variant`}</inlineCode>{` or `}<inlineCode parentName="p">{`NULL`}</inlineCode>{` if either of the arguments is `}<inlineCode parentName="p">{`NULL`}</inlineCode>{`.`}</p>
10:23:22.794 | > 10 \| <p><inlineCode parentName="p">{`GET_PATH`}</inlineCode>{` is equivalent to a chain of `}<inlineCode parentName="p">{`GET`}</inlineCode>{` functions, `}<inlineCode parentName="p">{`path_name`}</inlineCode>{` consists of a concatenation of field names preceded by periods (.), colons (:) or index operators (`}{`[`}<index>{`]`}{`). The first field name does not require the leading identifier to be specified.`}</p>
10:23:22.794 | \|                                                                                                                                                                                                                                                                                                                                                                                                                          ^
10:23:22.794 | 11 \| <h2 {...{"id":"syntax"}}>{`Syntax`}</h2>
10:23:22.794 | 12 \| <pre><code parentName="pre" {...{"className":"language-sql"}}>{`get_path(array, path_name)
10:23:22.794 | 13 \| get_path(object, path_name)
10:23:22.795 | [ERROR] Client bundle compiled with errors therefore further build is impossible.
10:23:22.843 | error Command failed with exit code 1.
10:23:22.844 | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
10:23:22.852 | Error: Command "yarn run build" exited with 1


```

**Comment by b41sh** [comment]
> ```
> 
> 
> SyntaxError: /vercel/path0/docs/doc/30-reference/20-functions/110-semi-structured-functions/get_path.md: Expected corresponding JSX closing tag for <index>. (10:409)
> --
> 10:23:22.794 | 8 \| <p>{`Extracts value from an `}<inlineCode parentName="p">{`ARRAY`}</inlineCode>{`, an `}<inlineCode parentName="p">{`OBJECT`}</inlineCode>{`, or a `}<inlineCode parentName="p">{`VARIANT`}</inlineCode>{` by `}<inlineCode parentName="p">{`path_name`}</inlineCode>{`.
> 10:23:22.794 | 9 \| The value is returned as a `}<inlineCode parentName="p">{`Variant`}</inlineCode>{` or `}<inlineCode parentName="p">{`NULL`}</inlineCode>{` if either of the arguments is `}<inlineCode parentName="p">{`NULL`}</inlineCode>{`.`}</p>
> 10:23:22.794 | > 10 \| <p><inlineCode parentName="p">{`GET_PATH`}</inlineCode>{` is equivalent to a chain of `}<inlineCode parentName="p">{`GET`}</inlineCode>{` functions, `}<inlineCode parentName="p">{`path_name`}</inlineCode>{` consists of a concatenation of field names preceded by periods (.), colons (:) or index operators (`}{`[`}<index>{`]`}{`). The first field name does not require the leading identifier to be specified.`}</p>
> 10:23:22.794 | \|                                                                                                                                                                                                                                                                                                                                                                                                                          ^
> 10:23:22.794 | 11 \| <h2 {...{"id":"syntax"}}>{`Syntax`}</h2>
> 10:23:22.794 | 12 \| <pre><code parentName="pre" {...{"className":"language-sql"}}>{`get_path(array, path_name)
> 10:23:22.794 | 13 \| get_path(object, path_name)
> 10:23:22.795 | [ERROR] Client bundle compiled with errors therefore further build is impossible.
> 10:23:22.843 | error Command failed with exit code 1.
> 10:23:22.844 | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
> 10:23:22.852 | Error: Command "yarn run build" exited with 1
> ```

done.

File: common/functions/src/scalars/semi_structureds/get.rs (hunk 1 above)
Comment (b41sh) [reply common/functions/src/scalars/semi_structureds/get.rs]: done.

File: common/functions/src/scalars/semi_structureds/get.rs [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    fn name(&self) -> &str {
+        &*self.display_name
+    }
+
+    fn return_type(&self, args: &[&DataTypePtr]) -> Result<DataTypePtr> {
+        let data_type = remove_nullable(args[0]);
+        let path_type = remove_nullable(args[1]);
+
+        if (IGNORE_CASE
+            && ((!data_type.is_null() && !data_type.data_type_id().is_variant_or_object())
```
Comment (sundy-li) [inline common/functions/src/scalars/semi_structureds/get.rs:67]: We really don't need to care about the null. Let's make it cleaner and more simple.

File: common/functions/src/scalars/semi_structureds/get.rs [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+        }
+        let path_keys = if BY_PATH {
+            parse_path_keys(columns[1].column())?
+        } else {
+            build_path_keys(columns[1].column())?
+        };
+
+        let data_type = columns[0].field().data_type();
+        if data_type.is_null() {
+            return NullType::arc().create_constant_column(&DataValue::Null, input_rows);
```
Comment (sundy-li) [inline common/functions/src/scalars/semi_structureds/get.rs:104]: Ditto

File: common/functions/src/scalars/semi_structureds/get.rs [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    let mut builder = NullableColumnBuilder::<JsonValue>::with_capacity(input_rows);
+    for path_key in path_keys.iter() {
+        if path_key.is_empty() {
+            for _ in 0..column.len() {
+                builder.append_null();
+            }
+            continue;
+        }
+        for v in column.iter() {
+            let mut found_value = true;
```
Comment (sundy-li) [inline common/functions/src/scalars/semi_structureds/get.rs:196]: Will found_value defaults to false make this more simple?

**Review by sundy-li — APPROVED** [review]

File: common/functions/src/scalars/semi_structureds/get.rs (hunk 4 above)
Comment (b41sh) [reply common/functions/src/scalars/semi_structureds/get.rs:196]: ok

**Comment by BohuTANG** [comment]
```
+++ /workspace/tests/suites/0_stateless/02_function/02_0051_function_semi_structureds_get.stdout	2022-04-07 05:02:22.259013252 +0000
@@ -1,3 +1,18 @@
+ERROR 2013 (HY000) at line 33: Lost connection to MySQL server during query
+ERROR 2006 (HY000) at line 34: MySQL server has gone away
+ERROR 2006 (HY000) at line 35: MySQL server has gone away
+ERROR 2006 (HY000) at line 36: MySQL server has gone away
+ERROR 2006 (HY000) at line 38: MySQL server has gone away
+ERROR 2006 (HY000) at line 39: MySQL server has gone away
+ERROR 2006 (HY000) at line 40: MySQL server has gone away
+ERROR 2006 (HY000) at line 42: MySQL server has gone away
+ERROR 2006 (HY000) at line 43: MySQL server has gone away
+ERROR 2006 (HY000) at line 44: MySQL server has gone away
+ERROR 2006 (HY000) at line 45: MySQL server has gone away
+ERROR 2006 (HY000) at line 46: MySQL server has gone away
+ERROR 2006 (HY000) at line 47: MySQL server has gone away
+ERROR 2006 (HY000) at line 48: MySQL server has gone away
+ERROR 2006 (HY000) at line 50: MySQL server has gone away
 ==get==
 2.71
 NULL
@@ -13,18 +28,4 @@
 "databend"
 "ext"
 NULL
-+==get from table==
-1
-NULL
-NULL
-1
-==get_ignore_case from table==
-1
-1
-==get_path from table==
-1
-"a"
-1
-1
-2
```

**Comment by b41sh** [comment]
> ```
> +++ /workspace/tests/suites/0_stateless/02_function/02_0051_function_semi_structureds_get.stdout	2022-04-07 05:02:22.259013252 +0000
> @@ -1,3 +1,18 @@
> +ERROR 2013 (HY000) at line 33: Lost connection to MySQL server during query
> +ERROR 2006 (HY000) at line 34: MySQL server has gone away
> +ERROR 2006 (HY000) at line 35: MySQL server has gone away
> +ERROR 2006 (HY000) at line 36: MySQL server has gone away
> +ERROR 2006 (HY000) at line 38: MySQL server has gone away
> +ERROR 2006 (HY000) at line 39: MySQL server has gone away
> +ERROR 2006 (HY000) at line 40: MySQL server has gone away
> +ERROR 2006 (HY000) at line 42: MySQL server has gone away
> +ERROR 2006 (HY000) at line 43: MySQL server has gone away
> +ERROR 2006 (HY000) at line 44: MySQL server has gone away
> +ERROR 2006 (HY000) at line 45: MySQL server has gone away
> +ERROR 2006 (HY000) at line 46: MySQL server has gone away
> +ERROR 2006 (HY000) at line 47: MySQL server has gone away
> +ERROR 2006 (HY000) at line 48: MySQL server has gone away
> +ERROR 2006 (HY000) at line 50: MySQL server has gone away
>  ==get==
>  2.71
>  NULL
> @@ -13,18 +28,4 @@
>  "databend"
>  "ext"
>  NULL
> -+==get from table==
> -1
> -NULL
> -NULL
> -1
> -==get_ignore_case from table==
> -1
> -1
> -==get_path from table==
> -1
> -"a"
> -1
> -1
> -2
> ```

this is caused by another bug, I will fix it.

**Comment by BohuTANG** [comment]
Conflicts that must be resolved :D

**Pull request merged**